File: frameworks/webonyx-graphql-php/tests/TestFactory.php
<?php

declare(strict_types=1);

namespace VelocityBench\Tests;

use DateTime;
use DateTimeImmutable;
use Ramsey\Uuid\Uuid;

/**
 * Test data structures
 */
class TestUser
{
    public string $id;
    public int $pkUser;
    public string $username;
    public string $email;
    public string $fullName;
    public ?string $bio;
    public DateTimeImmutable $createdAt;
    public DateTimeImmutable $updatedAt;

    public function __construct(
        string $id,
        int $pkUser,
        string $username,
        string $email,
        string $fullName,
        ?string $bio = null
    ) {
        $this->id = $id;
        $this->pkUser = $pkUser;
        $this->username = $username;
        $this->email = $email;
        $this->fullName = $fullName;
        $this->bio = $bio;
        $this->createdAt = new DateTimeImmutable();
        $this->updatedAt = new DateTimeImmutable();
    }
}

class TestPost
{
    public string $id;
    public int $pkPost;
    public int $fkAuthor;
    public string $authorId;
    public string $title;
    public string $content;
    public DateTimeImmutable $createdAt;
    public DateTimeImmutable $updatedAt;
    public ?TestUser $author;

    public function __construct(
        string $id,
        int $pkPost,
        int $fkAuthor,
        string $authorId,
        string $title,
        string $content,
        ?TestUser $author = null
    ) {
        $this->id = $id;
        $this->pkPost = $pkPost;
        $this->fkAuthor = $fkAuthor;
        $this->authorId = $authorId;
        $this->title = $title;
        $this->content = $content;
        $this->author = $author;
        $this->createdAt = new DateTimeImmutable();
        $this->updatedAt = new DateTimeImmutable();
    }
}

class TestComment
{
    public string $id;
    public int $pkComment;
    public int $fkPost;
    public int $fkAuthor;
    public string $postId;
    public string $authorId;
    public string $content;
    public DateTimeImmutable $createdAt;
    public ?TestUser $author;
    public ?TestPost $post;

    public function __construct(
        string $id,
        int $pkComment,
        int $fkPost,
        int $fkAuthor,
        string $postId,
        string $authorId,
        string $content,
        ?TestUser $author = null,
        ?TestPost $post = null
    ) {
        $this->id = $id;
        $this->pkComment = $pkComment;
        $this->fkPost = $fkPost;
        $this->fkAuthor = $fkAuthor;
        $this->postId = $postId;
        $this->authorId = $authorId;
        $this->content = $content;
        $this->author = $author;
        $this->post = $post;
        $this->createdAt = new DateTimeImmutable();
    }
}

/**
 * In-memory test factory for isolated tests
 */
class TestFactory
{
    /** @var array<string, TestUser> */
    private array $users = [];

    /** @var array<string, TestPost> */
    private array $posts = [];

    /** @var array<string, TestComment> */
    private array $comments = [];

    private int $userCounter = 0;
    private int $postCounter = 0;
    private int $commentCounter = 0;

    public function createUser(
        string $username,
        string $email,
        string $fullName = 'Test User',
        ?string $bio = null
    ): TestUser {
        if (empty($username)) {
            throw new \InvalidArgumentException('Username cannot be empty');
        }
        if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
            throw new \InvalidArgumentException("Invalid email address: {$email}");
        }
        if (strlen($fullName) > 255) {
            throw new \InvalidArgumentException('Full name exceeds 255 character limit');
        }
        if ($bio !== null && strlen($bio) > 1000) {
            throw new \InvalidArgumentException('Bio exceeds 1000 character limit');
        }
        foreach ($this->users as $existing) {
            if ($existing->username === $username) {
                throw new \RuntimeException("Username already taken: {$username}");
            }
            if (isset($existing->email) && $existing->email === $email) {
                throw new \RuntimeException("Email already in use: {$email}");
            }
        }

        $this->userCounter++;
        $user = new TestUser(
            Uuid::uuid4()->toString(),
            $this->userCounter,
            $username,
            $email,
            $fullName,
            $bio
        );
        $this->users[$user->id] = $user;
        return $user;
    }

    public function createPost(
        string $authorId,
        string $title,
        string $content = 'Default content'
    ): TestPost {
        $author = $this->users[$authorId] ?? null;
        if ($author === null) {
            throw new \RuntimeException("Author not found: {$authorId}");
        }
        if (empty($title)) {
            throw new \InvalidArgumentException('Title cannot be empty');
        }
        if (strlen($title) > 255) {
            throw new \InvalidArgumentException('Title exceeds 255 character limit');
        }

        $this->postCounter++;
        $post = new TestPost(
            Uuid::uuid4()->toString(),
            $this->postCounter,
            $author->pkUser,
            $author->id,
            $title,
            $content,
            $author
        );
        $this->posts[$post->id] = $post;
        return $post;
    }

    public function createComment(
        string $postId,
        string $authorId,
        string $content
    ): TestComment {
        $author = $this->users[$authorId] ?? null;
        $post = $this->posts[$postId] ?? null;

        if ($author === null) {
            throw new \RuntimeException("Author not found");
        }
        if ($post === null) {
            throw new \RuntimeException("Post not found");
        }
        if (empty($content)) {
            throw new \InvalidArgumentException('Comment content cannot be empty');
        }

        $this->commentCounter++;
        $comment = new TestComment(
            Uuid::uuid4()->toString(),
            $this->commentCounter,
            $post->pkPost,
            $author->pkUser,
            $post->id,
            $author->id,
            $content,
            $author,
            $post
        );
        $this->comments[$comment->id] = $comment;
        return $comment;
    }

    public function getUser(string $id): ?TestUser
    {
        return $this->users[$id] ?? null;
    }

    public function getPost(string $id): ?TestPost
    {
        return $this->posts[$id] ?? null;
    }

    public function getComment(string $id): ?TestComment
    {
        return $this->comments[$id] ?? null;
    }

    /**
     * @return TestUser[]
     */
    public function getAllUsers(): array
    {
        return array_values($this->users);
    }

    /**
     * @return TestPost[]
     */
    public function getPostsByAuthor(int $authorPk): array
    {
        return array_values(array_filter(
            $this->posts,
            fn(TestPost $p) => $p->fkAuthor === $authorPk
        ));
    }

    /**
     * @return TestComment[]
     */
    public function getCommentsByPost(int $postPk): array
    {
        return array_values(array_filter(
            $this->comments,
            fn(TestComment $c) => $c->fkPost === $postPk
        ));
    }

    /**
     * @return TestPost[]
     */
    public function getAllPosts(): array
    {
        return array_values($this->posts);
    }

    public function updateUser(string $id, ?string $bio, string $fullName = 'Test User'): ?TestUser
    {
        if (!isset($this->users[$id])) {
            return null;
        }
        $this->users[$id]->bio = $bio;
        $this->users[$id]->fullName = $fullName;
        $this->users[$id]->updatedAt = new DateTimeImmutable();
        return $this->users[$id];
    }

    public function updatePost(string $id, string $title, string $content): ?TestPost
    {
        if (!isset($this->posts[$id])) {
            return null;
        }
        $this->posts[$id]->title = $title;
        $this->posts[$id]->content = $content;
        $this->posts[$id]->updatedAt = new DateTimeImmutable();
        return $this->posts[$id];
    }

    public function deleteUser(string $id): void
    {
        unset($this->users[$id]);
    }

    public function deletePost(string $id): void
    {
        unset($this->posts[$id]);
    }

    public function reset(): void
    {
        $this->users = [];
        $this->posts = [];
        $this->comments = [];
        $this->userCounter = 0;
        $this->postCounter = 0;
        $this->commentCounter = 0;
    }
}

class ValidationHelper
{
    private const UUID_REGEX = '/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i';

    public static function isValidUuid(string $value): bool
    {
        return preg_match(self::UUID_REGEX, $value) === 1;
    }
}

class DataGenerator
{
    public static function generateLongString(int $length): string
    {
        return str_repeat('x', $length);
    }

    public static function generateRandomUsername(): string
    {
        return 'user_' . bin2hex(random_bytes(4));
    }
}

File: frameworks/hasura/tests/test_factory.py
"""In-memory test factory for Hasura integration tests."""

import uuid
from datetime import datetime, timezone
from dataclasses import dataclass, field


@dataclass
class TestUser:
    id: str
    pk_user: int
    username: str
    full_name: str
    bio: str | None
    created_at: datetime
    updated_at: datetime


@dataclass
class TestPost:
    id: str
    pk_post: int
    fk_author: int
    title: str
    content: str
    created_at: datetime
    updated_at: datetime
    author: TestUser | None = None


@dataclass
class TestComment:
    id: str
    pk_comment: int
    fk_post: int
    fk_author: int
    content: str
    created_at: datetime
    author: TestUser | None = None
    post: TestPost | None = None


class TestFactory:
    """In-memory test factory for isolated tests."""

    def __init__(self):
        self.users: dict[str, TestUser] = {}
        self.posts: dict[str, TestPost] = {}
        self.comments: dict[str, TestComment] = {}
        self._user_counter = 0
        self._post_counter = 0
        self._comment_counter = 0

    def create_user(
        self,
        username: str,
        email: str,
        full_name: str,
        bio: Optional[str] = None
    ) -> TestUser:
        self._user_counter += 1
        user = TestUser(
            id=str(uuid.uuid4()),
            pk_user=self._user_counter,
            username=username,
            full_name=full_name,
            bio=bio,
            created_at=datetime.now(timezone.utc),
            updated_at=datetime.now(timezone.utc),
        )
        self.users[user.id] = user
        return user

    def create_post(
        self,
        author_id: str,
        title: str,
        content: str = "Default content"
    ) -> TestPost:
        author = self.users.get(author_id)
        if not author:
            raise ValueError(f"Author not found: {author_id}")

        self._post_counter += 1
        post = TestPost(
            id=str(uuid.uuid4()),
            pk_post=self._post_counter,
            fk_author=author.pk_user,
            title=title,
            content=content,
            created_at=datetime.now(timezone.utc),
            updated_at=datetime.now(timezone.utc),
            author=author,
        )
        self.posts[post.id] = post
        return post

    def create_comment(
        self,
        author_id: str,
        post_id: str,
        content: str
    ) -> TestComment:
        author = self.users.get(author_id)
        post = self.posts.get(post_id)

        if not author:
            raise ValueError("Author not found")
        if not post:
            raise ValueError("Post not found")

        self._comment_counter += 1
        comment = TestComment(
            id=str(uuid.uuid4()),
            pk_comment=self._comment_counter,
            fk_post=post.pk_post,
            fk_author=author.pk_user,
            content=content,
            created_at=datetime.now(timezone.utc),
            author=author,
            post=post,
        )
        self.comments[comment.id] = comment
        return comment

    def get_user(self, id: str) -> Optional[TestUser]:
        return self.users.get(id)

    def get_post(self, id: str) -> Optional[TestPost]:
        return self.posts.get(id)

    def get_comment(self, id: str) -> Optional[TestComment]:
        return self.comments.get(id)

    def get_all_users(self) -> list[TestUser]:
        return list(self.users.values())

    def get_posts_by_author(self, author_pk: int) -> list[TestPost]:
        return [p for p in self.posts.values() if p.fk_author == author_pk]

    def get_comments_by_post(self, post_pk: int) -> list[TestComment]:
        return [c for c in self.comments.values() if c.fk_post == post_pk]

    def reset(self):
        self.users.clear()
        self.posts.clear()
        self.comments.clear()
        self._user_counter = 0
        self._post_counter = 0
        self._comment_counter = 0


class ValidationHelper:
    UUID_REGEX = r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"

    @staticmethod
    def is_valid_uuid(value: str) -> bool:
        import re
        return bool(re.match(ValidationHelper.UUID_REGEX, value.lower()))


class DataGenerator:
    @staticmethod
    def generate_long_string(length: int) -> str:
        return "x" * length

    @staticmethod
    def generate_random_username() -> str:
        return f"user_{uuid.uuid4().hex[:8]}"
